lib/sms/sms_deliver.c

```diff
@@ -0,0 +1,670 @@
+{
+	/* 4 LSBs represent decimal that should be multiplied by 10. */
+	/* 4 MSBs represent decimal that should be add as is. */
+	return ((value & 0xf0) >> 4) + ((value & 0x0f) * 10);
+}
+
+static int decode_pdu_deliver_header(struct parser *parser, uint8_t *buf)
+{
+	uint8_t smsc_size = *buf;
+	buf += smsc_size + 1;
+
+	LOG_DBG("SMSC size: %d", smsc_size);
+
+	DELIVER_DATA(parser)->field_header = 
+		*((struct pdu_deliver_header*)buf);
+
+	LOG_DBG("SMS header 1st byte: 0x%X", *buf);
+
+	LOG_DBG("TP-Message-Type-Indicator: %d", DELIVER_DATA(parser)->field_header.mti);
+	LOG_DBG("TP-More-Messages-to-Send: %d", DELIVER_DATA(parser)->field_header.mms);
+	LOG_DBG("TP-Status-Report-Indication: %d", DELIVER_DATA(parser)->field_header.sri);
+	LOG_DBG("TP-User-Data-Header-Indicator: %d", DELIVER_DATA(parser)->field_header.udhi);
+	LOG_DBG("TP-Reply-Path: %d", DELIVER_DATA(parser)->field_header.rp);
+
+	return smsc_size + 2;
+}
+
+static int decode_pdu_do_field(struct parser *parser, uint8_t *buf)
+{
+
+	DELIVER_DATA(parser)->field_oa.length   = (uint8_t)*buf++;
+	DELIVER_DATA(parser)->field_oa.adr_type = (uint8_t)*buf++;
+
+	LOG_DBG("Address-Length: %d", DELIVER_DATA(parser)->field_oa.length);
+	LOG_DBG("Type-of-Address: %02X", DELIVER_DATA(parser)->field_oa.adr_type);
+
+	if (DELIVER_DATA(parser)->field_oa.length > SMS_MAX_ADDRESS_LEN_CHARS) {
+		LOG_ERR("Maximum address length (%d) exceeded %d. Aborting decoding.",
+			SMS_MAX_ADDRESS_LEN_OCTETS,
+			DELIVER_DATA(parser)->field_oa.length);
+		return -EINVAL;
+	}
+
+	uint8_t length = DELIVER_DATA(parser)->field_oa.length / 2;
+	if (DELIVER_DATA(parser)->field_oa.length % 2 == 1) {
+		/* There is an extra number in semi-octet and fill bits*/
+		length++;
+	}
+
+	memcpy(DELIVER_DATA(parser)->field_oa.adr,
+	       buf, 
+	       length);
+
+	for(int i = 0; i < length; i++) {
+		DELIVER_DATA(parser)->field_oa.adr[i] = 
+			swap_nibbles(DELIVER_DATA(parser)->field_oa.adr[i]);
+	}
+
+	return 2 + length;
+}
+
+static int decode_pdu_pid_field(struct parser *parser, uint8_t *buf)
+{
+	DELIVER_DATA(parser)->field_pid = (uint8_t)*buf;
+
+	LOG_DBG("TP-Protocol-Identifier: %d", DELIVER_DATA(parser)->field_pid);
+
+	return 1;
+}
+
+static int decode_pdu_dcs_field(struct parser *parser, uint8_t *buf)
+{
+	DELIVER_DATA(parser)->field_dcs = *((struct pdu_dcs_field*)buf);
+
+	LOG_DBG("TP-Data-Coding-Scheme: %02X", *buf);
+
+	return 1;
+}
+
+static int decode_pdu_scts_field(struct parser *parser, uint8_t *buf)
+{
+	int tmp_tz;
+	
+	DELIVER_DATA(parser)->timestamp.year   = semioctet_to_dec(*(buf++));
+	DELIVER_DATA(parser)->timestamp.month  = semioctet_to_dec(*(buf++));
+	DELIVER_DATA(parser)->timestamp.day    = semioctet_to_dec(*(buf++));
+	DELIVER_DATA(parser)->timestamp.hour   = semioctet_to_dec(*(buf++));
+	DELIVER_DATA(parser)->timestamp.minute = semioctet_to_dec(*(buf++));
+	DELIVER_DATA(parser)->timestamp.second = semioctet_to_dec(*(buf++));
+
+	tmp_tz = ((*buf&0xf7) * 15) / 60;
```

Comment:
Interesting bit magic, can this function (or this line at least) have more comments? :)

Reply:
Comments added by copying 3GPP definition for timezone.